File: leetcode/0203-remove-linked-list-elements/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        """
        RUNTIME: 66 ms (96.74%)
        MEMORY: 17.8 MB (83.99%)
        """
        while head and head.val == val:
            tmp = head.next
            head = tmp
        if not head: return
        current = head
        while current.next:
            if current.next.val == val:
                current.next = current.next.next
            else:
                current = current.next
        return head

File: leetcode/0203-remove-linked-list-elements/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    /**
     * RUNTIME: 31 ms (73.75%)
     * MEMORY: 15 MB (81.27%)
    */
    ListNode* removeElements(ListNode* head, int val) {
        ListNode *tmp;
        while (head && head->val == val) {
            tmp = head->next;
            head = tmp;
        }
        if (!head) { return NULL; }
        ListNode *current = head;
        while (current->next) {
            if (current->next->val == val) {
                current->next = current->next->next;
            }
            else {
                current = current->next;
            }
        }
        return head;
    }
};